File: 102-binary-tree-level-order-traversal/binary-tree-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode*  root) {
    vector<vector<int>> res; 
    vector<int> ans;
    if (root == NULL) {
        return res;
    }
    queue<TreeNode*> q;
    q.push(root);
    q.push(NULL); // Marker for the end of the current level
    
    while (!q.empty()) {
        TreeNode* front = q.front();
        q.pop();
        if (front == NULL) {
            // End of the current level
            res.push_back(ans);
            ans.clear();
            // If there are more nodes to process, add a new level marker
            if (!q.empty()) {
                q.push(NULL);
            }
        } else {
            ans.push_back(front->val);
            if (front->left != NULL) {
                q.push(front->left);
            }
            if (front->right != NULL) {
                q.push(front->right);
            }
        }
    }

    return res;
}
};

File: 102-binary-tree-level-order-traversal/binary-tree-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque
class Solution(object):
    def levelOrder(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: List[List[int]]
        """
        result = []
        # Handle empty tree case
        if root is None:
            return result
      
        # Initialize queue with root node for BFS
        q = deque([root])
      
        # Process tree level by level
        while q:
            current_level = []
            level_size = len(q)
          
            # Process all nodes at the current level
            for _ in range(level_size):
                # Remove node from front of queue
                node = q.popleft()
              
                # Add node value to current level list
                current_level.append(node.val)
              
                # Add children to queue for next level processing
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
          
            # Add current level to result
            result.append(current_level)
      
        return result



        
        